File: CatInCup/src/com/android101/example/catincup/MainActivity.java
package com.android101.example.catincup;

/* CatInCup is a game where the user tries to guess in which of the cups a cat is hiding.
 * The user starts with 9 lives and his/her goal is to remain alive for as many rounds as possible.
 * */

import android.os.Bundle;
import android.app.Activity;
import android.app.AlertDialog;
import android.content.DialogInterface;
import android.view.Menu;
import android.widget.ImageButton;
import android.widget.TextView;
//import android.util.Log;
import android.view.View;
import android.view.View.OnClickListener;
import java.util.Random;

public class MainActivity extends Activity {
	private final int MAX_CUPS = 3;
	private final int ACTION_HIT = 0;
	private final int ACTION_MISS = 1;
	private final int ACTION_NONE = 3;
    private Random generator;
    private int lives, rounds, lastAction;
    private TextView livesView;
    private TextView roundsView;
    private ImageButton []b;
    
    // when the user changes the position of the device between land and port,
    // android creates a new Activity, so we must make sure to preserve 
    // the game score. 
    @Override
    public void onSaveInstanceState(Bundle savedInstanceState) {
        savedInstanceState.putInt("CurrentLives", lives);
        savedInstanceState.putInt("CurrentRounds", rounds);
        savedInstanceState.putInt("LastAction", lastAction);
        super.onSaveInstanceState(savedInstanceState);
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        
        // force the orientation to be landscape from the start
        setRequestedOrientation(android.content.pm.ActivityInfo.SCREEN_ORIENTATION_LANDSCAPE);
        
        // set up the random number generator
        generator = new Random(System.currentTimeMillis());
        
        // save pointers to the frequently-used widgets
        livesView = ((TextView) findViewById(R.id.TextViewLives));
        roundsView = ((TextView) findViewById(R.id.TextViewRounds));
        
        b = new ImageButton[3];
        b[0] = ((ImageButton) findViewById(R.id.ImageButtonCup1));
        b[1] = ((ImageButton) findViewById(R.id.ImageButtonCup2));
        b[2] = ((ImageButton) findViewById(R.id.ImageButtonCup3));
        
        // set up the event handler for the buttons
        b[0].setOnClickListener(click);
        b[1].setOnClickListener(click);
        b[2].setOnClickListener(click);
        
        // setting game-play variables and interfaces for the first time
        if(savedInstanceState == null) {
        	start();
        	
        // propagating an existing state
        } else {
	        super.onRestoreInstanceState(savedInstanceState);
	        lives = savedInstanceState.getInt("CurrentLives");
	        rounds = savedInstanceState.getInt("CurrentRounds");
	        lastAction = savedInstanceState.getInt("LastAction");
	        refreshScore(lastAction);
        }
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        // Inflate the menu; this adds items to the action bar if it is present.
        getMenuInflater().inflate(R.menu.main, menu);
        return true;
    }
    
    public void refreshScore(int action) {
    	if(action == ACTION_NONE) {
    		livesView.setTextColor(getResources().getColor(R.color.none_color));
    	} else if(action == ACTION_HIT) {
    		livesView.setTextColor(getResources().getColor(R.color.hit_color));
    	} else if(action == ACTION_MISS) {
    		livesView.setTextColor(getResources().getColor(R.color.miss_color));
    	}
    	lastAction = action;
    	livesView.setText(getString(R.string.lives) + " " + lives);
    	roundsView.setText(getString(R.string.rounds) + " " + rounds);
    }
    
    public void start() {
    	lives = 9;
    	rounds = 1;
    	refreshScore(ACTION_NONE);
    	clear_cups();
    }

    public void clear_cups() {
    	b[0].setImageResource(R.drawable.cup_empty);
    	b[1].setImageResource(R.drawable.cup_empty);
    	b[2].setImageResource(R.drawable.cup_empty);
    }
    
    public void show_cat(boolean win, ImageButton b) {
    	if(win) // player wins round
    	  b.setImageResource(R.drawable.cup_win1);
    	else // player loses round
      	  b.setImageResource(R.drawable.cup_lose1);
    	
    	new android.os.Handler().postDelayed(new Runnable() {
            @Override
            public void run() {
            	clear_cups();
            }
        }, 700);
    	
    	//Log.v("CATINCUP","User guessed " + (win ? "right" : "wrong"));
    }
    
    public void evaluate_guess(int guess) {
    	 int cup = (generator.nextInt() % MAX_CUPS);
    	 if(cup < 0) cup = -cup;
    	 //Log.v("CATINCUP","CUP: " + cup);

    	 if(cup == guess-1) {
    		// user guesses right
    		 lives++;
    		 show_cat(true, b[cup]);
    		 refreshScore(ACTION_HIT);
    		 
    	 } else {
    		// user guesses wrong
    		 lives--;
    		 show_cat(false, b[cup]);
    		 refreshScore(ACTION_MISS);
    	 }
    	 
    	 // did we run out of lives?
    	 if(lives == 0) {
    		 game_over();
    	 } else {
        	 rounds++;
    	 }
    }
    
    public void game_over() {
    	// create dialogue to inform end of game and ask if user wants to continue
		AlertDialog.Builder alertDialogBuilder = new AlertDialog.Builder(this);

		// set up the dialog's properties (using a factory class)
		alertDialogBuilder //.setMessage("Click New Game to continue playing!")
			.setTitle("You survived for " + rounds + " rounds! Play again?")
			.setCancelable(false)
			.setPositiveButton("New Game",new DialogInterface.OnClickListener() {
				public void onClick(DialogInterface dialog,int id) {
					start();
				}
			})
			.setNegativeButton("Quit",new DialogInterface.OnClickListener() {
				public void onClick(DialogInterface dialog,int id) {
					finish(); //dialog.cancel();
				}
			});
 
		// create and show the dialog box
		AlertDialog alertDialog = alertDialogBuilder.create();
		alertDialog.show();    	
    }
    
    OnClickListener click = new OnClickListener() {
        @Override
        public void onClick(View v) {
        		if(v.getId() == R.id.ImageButtonCup1) {
                	evaluate_guess(1);
                } else if(v.getId() == R.id.ImageButtonCup2) {
                	evaluate_guess(2);
                } else if(v.getId() == R.id.ImageButtonCup3) {
                	evaluate_guess(3);
                }
        }
    };
}

File: CatsNCups/app/src/main/java/com/koalascent/catsncups/MainActivity.kt
package com.koalascent.catsncups

import android.app.AlertDialog
import android.os.Bundle
import android.os.Handler
import android.os.Vibrator
import android.view.View.*
import android.widget.ImageButton
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.ContextCompat
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProviders
import com.google.android.material.snackbar.Snackbar
import com.koalascent.catsncups.game.GameAction
import com.koalascent.catsncups.game.GameActionType
import com.koalascent.catsncups.game.GameState
import com.koalascent.catsncups.game.GameViewModel
import com.koalascent.catsncups.haptic.HapticFeedbackHelper
import kotlinx.android.extensions.CacheImplementation
import kotlinx.android.extensions.ContainerOptions
import kotlinx.android.synthetic.main.activity_main.*

@ContainerOptions(CacheImplementation.SPARSE_ARRAY)
class MainActivity : AppCompatActivity() {

    private lateinit var gameModel: GameViewModel
    private lateinit var livesView: TextView
    private lateinit var scoreView: TextView
    private lateinit var cupButtons: ArrayList<ImageButton>
    private lateinit var handler: Handler
    private lateinit var vibrator: Vibrator

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        livesView = findViewById(R.id.livesTextView)
        scoreView = findViewById(R.id.roundTextView)
        vibrator = HapticFeedbackHelper.vibratorFor(this)

        cupButtons = arrayListOf(
            findViewById<ImageButton>(R.id.cupButton1).also { it.setOnClickListener(click) },
            findViewById<ImageButton>(R.id.cupButton2).also { it.setOnClickListener(click) },
            findViewById<ImageButton>(R.id.cupButton3).also { it.setOnClickListener(click) }
        )

        handler = Handler()
        gameModel = ViewModelProviders.of(this).get(GameViewModel::class.java)

        gameModel.observeLives().observe(this, Observer { lives ->
            livesView.text = getString(R.string.lives, lives)
        })

        gameModel.observeScore().observe(this, Observer { score ->
            scoreView.text = getString(R.string.score, score)
        })

        gameModel.observeAction().observe(this, Observer { action ->
            performAction(action)
        })

        gameModel.observeGameState().observe(this, Observer { gameState ->
            if (gameState == GameState.STARTED) {
                showGameStart()
                resetCups()
                showCups()
                gameModel.nextMove()
            } else if (gameState == GameState.OVER) {
                hideCups()
                showGameOverDialog()
            }
        })
    }

    private val click: OnClickListener = OnClickListener { v ->
        val guessedCup = when (v.id) {
            R.id.cupButton1 -> 1
            R.id.cupButton2 -> 2
            R.id.cupButton3 -> 3
            else -> 0
        }
        if (guessedCup > 0) {
            handler.removeCallbacksAndMessages(null)
            resetCups()
            gameModel.evaluateGuess(guessedCup)
        }
    }

    private fun performAction(action: GameAction) {
        livesView.setTextColor(ContextCompat.getColor(this, livesColorFor(action)))

        when (action.type) {
            GameActionType.HIT -> {
                showCat(R.drawable.cup_win1, action.correctCup)
                HapticFeedbackHelper.vibrateFor(action.type, vibrator)
            }
            GameActionType.MISS -> {
                showCat(R.drawable.cup_lose1, action.correctCup)
                HapticFeedbackHelper.vibrateFor(action.type, vibrator)
            }
            else -> {
            }
        }
    }

    private fun livesColorFor(action: GameAction): Int = when (action.type) {
        GameActionType.HIT -> R.color.hitColor
        GameActionType.MISS -> R.color.missColor
        else -> R.color.noneColor
    }

    private fun showCat(imageResource: Int, cup: Int) {
        cupButtons[cup].let { selectedButton ->
            selectedButton.setImageResource(imageResource)
            handler.postDelayed({
                resetCups()
            }, 700)
        }
    }

    private fun showCups() {
        cupButtons.forEach { it.visibility = VISIBLE }
    }

    private fun hideCups() {
        cupButtons.forEach { it.visibility = INVISIBLE }
    }

    private fun resetCups() {
        cupButtons.forEach { it.setImageResource(R.drawable.cup_empty) }
    }

    private fun showGameStart() {
        Snackbar.make(rootLayout, R.string.instructions, Snackbar.LENGTH_INDEFINITE)
            .setAction(R.string.dismiss) {}
            .show()
    }

    // create dialogue to inform end of game and ask if user wants to continue
    private fun showGameOverDialog() {
        AlertDialog.Builder(this)
            .setTitle(getString(R.string.gameOver))
            .setMessage(getString(R.string.game_over_message, gameModel.observeScore().value))
            .setCancelable(false)
            .setPositiveButton(R.string.newGame) { _, _ -> gameModel.restartGame() }
            .setNegativeButton(R.string.quit) { _, _ -> finish() }
            .create()
            .show()
    }
}